controllers/discovery/initDiscoveryTopicData.js

```diff
@@ -19,41 +19,43 @@ const InitDiscoveryTopicData = async (req, res) => {
-                    *
-                    (0.2 + (SELECT 
-                    count(D.post_id) 
-                    FROM posts D 
-                    INNER JOIN post_topics E 
-                    ON D.post_id = E.post_id 
-                    INNER JOIN topics F 
-                    ON E.topic_id = F.topic_id 
-                    WHERE F.topic_id = A.topic_id 
-                    AND D.created_at > current_date - interval '7 days'
-                  ) ^ 0.5)) AS ordering_score
-              FROM user_topics A 
-              INNER JOIN user_topics B 
-                  ON A.topic_id = B.topic_id
-                  AND A.user_id in (:userId, :anonymousUserId)
-              RIGHT JOIN topics C 
-                  ON C.topic_id = A.topic_id
-              GROUP BY A.topic_id, C.topic_id, A.user_id
-              ORDER BY 
-                ordering_score DESC
-              LIMIT :limit
-              OFFSET :offset`;
+                          SELECT 
+                            C.*,
+                            COUNT(*) as common,
+                            A.user_id as user_id_follower,
+                            ((1 + 
+                              CASE
+                                WHEN (select count(*) from user_follow_user ufu where user_id_follower = :userId and user_id_followed in (
+                                    select user_id from user_topics ut where topic_id = C.topic_id
+                                  )) > 20 THEN 20
+                            ELSE 
+                                (select count(*) from user_follow_user ufu where user_id_follower = :userId and user_id_followed in (
+                                  select user_id from user_topics ut where topic_id = C.topic_id
+                                ))
+                            END
+                            )
+                              *
+                              (0.2 + (SELECT 
+                              count(D.post_id) 
+                              FROM posts D 
+                              INNER JOIN post_topics E 
+                              ON D.post_id = E.post_id 
+                              WHERE E.topic_id = A.topic_id 
+                              AND D.created_at > current_date - interval '7 days'
+                                ) ^ 0.5)) AS ordering_score
+                            FROM user_topics A 
+                            INNER JOIN user_topics B 
+                                ON A.topic_id = B.topic_id
+                                AND A.user_id in (:userId, :anonymousUserId)
+                            RIGHT JOIN topics C 
+                                ON C.topic_id = A.topic_id
+                            GROUP BY A.topic_id, C.topic_id, A.user_id
+                            ORDER BY 
+                              user_id_follower,
+                              ordering_score DESC
+                            LIMIT :limit
+                            OFFSET :offset`;
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

The new SQL query seems to be more complex than the previous one. It introduces subqueries that are executed multiple times, which could potentially impact performance. The subquery is used twice in the `CASE` statement (lines 29-31 and 33-35), which means it will be executed for each row in the result set. If the number of rows is large, this could lead to a significant slowdown.

Consider calculating the count of common followers once and storing it in a variable or as a column in a temporary table. This way, you can avoid executing the same subquery multiple times.

Also, the join condition on line 50 (`AND A.user_id in (:userId, :anonymousUserId)`) might not work as expected if either `:userId` or `:anonymousUserId` is null. In SQL, comparing anything with null using `=` or `in` results in null, which might exclude some rows from the result set. Consider handling null values explicitly.

Here's a simplified version of your query that addresses these issues:

```diff
- let suggestedTopicsQuery = `
- ...
- CASE
-   WHEN (select count(*) from user_follow_user ufu where user_id_follower = :userId and user_id_followed in (
-       select user_id from user_topics ut where topic_id = C.topic_id
-     )) > 20 THEN 20
- ELSE 
-   (select count(*) from user_follow_user ufu where user_id_follower = :userId and user_id_followed in (
-     select user_id from user_topics ut where topic_id = C.topic_id
-   ))
- END
- ...
- AND A.user_id in (:userId, :anonymousUserId)
- ...`;

+ let suggestedTopicsQuery = `
+ WITH follower_counts AS (
+   SELECT 
+     ut.topic_id,
+     COUNT(ufu.user_id_follower) AS follower_count
+   FROM user_follow_user ufu
+   INNER JOIN user_topics ut ON ufu.user_id_followed = ut.user_id
+   WHERE ufu.user_id_follower = :userId
+   GROUP BY ut.topic_id
+ )
+ SELECT 
+   C.*,
+   COUNT(*) as common,
+   A.user_id as user_id_follower,
+   ((1 + 
+     CASE
+       WHEN fc.follower_count > 20 THEN 20
+     ELSE 
+       fc.follower_count
+     END
+   )
+     *
+   (0.2 + (SELECT 
+     count(D.post_id) 
+     FROM posts D 
+     INNER JOIN post_topics E 
+     ON D.post_id = E.post_id 
+     WHERE E.topic_id = A.topic_id 
+     AND D.created_at > current_date - interval '7 days'
+   ) ^ 0.5)) AS ordering_score
+ FROM user_topics A 
+ INNER JOIN user_topics B 
+   ON A.topic_id = B.topic_id
+   AND (A.user_id = :userId OR (A.user_id IS NULL AND :userId IS NULL))
+   AND (A.user_id = :anonymousUserId OR (A.user_id IS NULL AND :anonymousUserId IS NULL))
+ RIGHT JOIN topics C 
+   ON C.topic_id = A.topic_id
+ LEFT JOIN follower_counts fc
+   ON fc.topic_id = C.topic_id
+ GROUP BY A.topic_id, C.topic_id, A.user_id
+ ORDER BY 
+   user_id_follower,
+   ordering_score DESC
+ LIMIT :limit
+ OFFSET :offset`;
```

This version of the query uses a Common Table Expression (CTE) to calculate the follower counts once and then uses that in the main query. It also handles null values in the join condition explicitly.


<!-- This is an auto-generated comment by OSS CodeRabbit -->